### PR TITLE
Use Buffer.MemoryCopy when available

### PR DESCRIPTION
**Pull request by benaadams**
MemoryCopy has improved a lot (and is faster than BlockCopy) https://github.com/dotnet/coreclr/pull/6627

And is undergoing further improvements https://github.com/dotnet/coreclr/pull/6638

### CODE REVIEW

**Comment by nietras** [comment]
Have you checked the benchmarks I did for .NET 4.6 RyuJIT 64-bit, not sure `MemoryCopy` is the best approach yet ;)
image

**Comment by benaadams** [comment]
Well its currently using BlockCopy so MemoryCopy is currently better :smile: 

And it looks like it will improve

**Comment by benaadams** [comment]
Updated https://github.com/aspnet/KestrelHttpServer/issues/1002 with further failure info

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs [hunk 1]
```diff
@@ -870,12 +870,30 @@ public int GetLength(MemoryPoolIterator end)
...
                 actual = 0;
                 return this;
             }
 
+            if (array == null)
+            {
+                return Consume(count, out actual);
+            }
+
+            Debug.Assert(count + offset <= array.Length);
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs]: Add message.

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs]: Maybe should change it to actual throw?

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs]: Also get the test the right way round XD

**Comment by benaadams** [comment]
I'd hold on this - movement in coreclr on BlockCopy https://github.com/dotnet/coreclr/pull/7198

**Comment by cesarbs** [comment]
@benaadams Any updates on this? https://github.com/dotnet/coreclr/pull/7198 was merged almost a month ago.

**Comment by benaadams** [comment]
@cesarbs that change is in post 1.1.0 😞 

Could take this now and add issue to remeasure post 1.1.0?

**Comment by halter73** [comment]
@benaadams Since this missed the 1.1 train and dotnet/coreclr#7198 is merged, should we close this?

**Comment by benaadams** [comment]
Yeah, think so - landscape has changed. Might be worth a revalidation at some point; but don't think its worth keeping this PR open

**Pull request closed without merging**